src/python/pants/backend/python/interpreter_cache.py

```diff
@@ -74,29 +74,34 @@ def _cache_dir(self):
 
   def select_interpreter_for_targets(self, targets):
     """Pick an interpreter compatible with all the specified targets."""
-    tgts_with_compatibilities = []
+    tgts_by_compatibilities = {}
     filters = set()
     for target in targets:
-      if isinstance(target, PythonTarget) and target.compatibility:
-        tgts_with_compatibilities.append(target)
-        filters.update(target.compatibility)
+      if isinstance(target, PythonTarget):
+        c = self._python_setup.compatibility_or_constraintstar(target)
+        targets = tgts_by_compatibilities.get(c)
+        if not targets:
+          targets = tgts_by_compatibilities[c] = []
+        targets.append(target)
+        filters.update(c)
 
     allowed_interpreters = set(self.setup(filters=filters))
```

Comment:
This call already follows the logic: if filters - use them, else use the `PythonSetup.interpreter_constraints` as the filters. Are you sure PIC needs modification?

Reply:
I am not sure that `PIC` needs modification, as I expect it was the edit in `python_execution_task_base.py` that had the effect.

But "if no targets specify `compatibility` values, then use the default" is not quite the same: if you have one target with no compatibility (that should thus use the default) and one target with, you should (attempt to) merge those.

This does make me wonder whether this logic needs graph awareness of some sort though.

Reply:
Agreed they're not the same, but this did read as a somewhat thrashy change trying a few things at once. It would be nice to revert without thorough tests if not needed.